appcues/src/main/java/com/appcues/qualification/DefaultQualificationService.kt

```diff
@@ -0,0 +1,241 @@
+package com.appcues.qualification
+
+import com.appcues.AppcuesConfig
+import com.appcues.analytics.AnalyticsIntent
+import com.appcues.analytics.AnalyticsIntent.Event
+import com.appcues.analytics.AnalyticsIntent.Identify
+import com.appcues.analytics.AnalyticsIntent.UpdateGroup
+import com.appcues.analytics.QualificationService
+import com.appcues.analytics.SdkMetrics
+import com.appcues.data.MoshiConfiguration
+import com.appcues.data.local.AppcuesLocalSource
+import com.appcues.data.local.model.ActivityStorage
+import com.appcues.data.mapper.experience.ExperienceMapper
+import com.appcues.data.model.ExperiencePriority
+import com.appcues.data.model.ExperiencePriority.LOW
+import com.appcues.data.model.ExperiencePriority.NORMAL
+import com.appcues.data.model.ExperienceTrigger
+import com.appcues.data.model.QualificationResult
+import com.appcues.data.remote.RemoteError.NetworkError
+import com.appcues.data.remote.appcues.AppcuesRemoteSource
+import com.appcues.data.remote.appcues.request.ActivityRequest
+import com.appcues.data.remote.appcues.request.EventRequest
+import com.appcues.data.session.PrefSessionLocalSource
+import com.appcues.logging.Logcues
+import com.appcues.util.doIfFailure
+import com.appcues.util.doIfSuccess
+import com.squareup.moshi.JsonDataException
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import java.util.Date
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+
+internal class DefaultQualificationService(
+    private val appcuesRemoteSource: AppcuesRemoteSource,
+    private val appcuesLocalSource: AppcuesLocalSource,
+    private val sessionLocalSource: PrefSessionLocalSource,
+    private val experienceMapper: ExperienceMapper,
+    private val config: AppcuesConfig,
+    private val logcues: Logcues,
+) : QualificationService {
+
+    private val processingActivity: HashSet<UUID> = hashSetOf()
+    private val mutex = Mutex()
+
+    override suspend fun qualify(intents: List<AnalyticsIntent>): QualificationResult? {
+        val activityRequest = intents.mergeToActivityRequest()
+
+        // FIXME trying to recall reason why old call site flushPendingActivity could accept null as date.
+        SdkMetrics.tracked(activityRequest.requestId, Date())
```

Comment:
if i recall, it was because _some_ network calls we know are not qualifying - just emptying the queue after 10 sec, like background flow analytics, and we didn't want to track metrics for those?

Reply:
ok I guess in that case it would be interesting to change how SdkMetrics work in a way that we figure out if we should track activityRequest or not, without depending on the timestamp value, then we can change this call to track(activityRequest) and not worry about loose logic around timestamp being null or not

Reply:
yeah i'm not 100% sure on previous usage either, need to investigate that one more - but surely it can change if we can find a more appropriate way to do it.

Reply:
need to figure out another way to filter out unwanted Activity requests  from SdkMetrics, or if we dont care we could just "track" all of them since its not a big lift, we just report those that come back with qualified experiences that can show == true anyway.